### PR TITLE
Schema type checker

### LINKED ISSUE #274: [Feature Request] Generate tests that can check against an actual postgres schema
It would be really nice to be able to be able to generate a test suite that will connect to a local postgres instance and verify that all `TableSchema` types match the postgres database's schema.

I am imagining that you would register each table schema in a list and maybe provide a pg connection string or something, and then the tests would  be generated.

I am happy to help implement this, but I would definitely need lots of direction.

### PR DESCRIPTION
**Pull request by abigailalice**
This should close #274 and #186, and potentially #291. The semantics are primarily based on @KaneTW's gist in #274, except with two improvements:

1. It handles array types
2. It returns an error if two columns in a `TableSchema` will have identical names, instead of passing silently

It also displays information in a slightly different way, trying to show as many errors as it can rather than bailing on the first. It also adds a number of test cases to ensure it behaves correctly for the primitive types, higher kinded types, and a few tests to ensure it fails in cases where it should fail. I'm mostly putting this forward to see what other people think, as I'm not sure if it should have other changes before being accepted. Some of my (admittedly bikesheddy) thoughts on the current state:

1. I wasn't able to re-export it from `Rel8` without creating an import cycle, so it's only exported via its own module being public. For some reason the module wouldn't compile unless it imported `Rel8`, getting a type error that `Columns` wasn't injective otherwise. I'm assuming the issue is that something not being in scope is effecting type checking, but I've never actually seen an error like that, so really aren't sure. I'm also not really sure if `Rel8.Table.Verify` is the best name for the module, I just picked `Table` pretty arbitrarily.
2. I reused `tests/Rel8/Generic/Rel8able/Test.hs` quite a bit to add some things necessary for my own test cases, but I'm not really sure if that's an appropriate place for them. I could move them if that's preferred.
3. The current work won't show KaneTW's original work, since I just copy-pasted it. Not sure if there's a proper way to base work off a gist, or if it's even an issue, but I thought it should still be said
4. I kind of want to have a version of the type checking function which also reads the data from the table after checking the schema, as having a valid schema isn't enough to ensure existing data can be read by rel8, which can be important when doing migrations. However, this isn't possible just using the `Statement` type. I just add a version using `Session` which does this, but for people using `hasql-transaction` they would still need to rewrite that function manually. This isn't a terribly difficult thing for users of the library to do themselves, since anyone using it is already going to need to convert from `Statement` to `Transaction`, and can just read from the table after type checking, but it just feels like an oversite to not have this. For the moment the issue is explained in the documentation for the function.

### CODE REVIEW

**Comment by KaneTW** [comment]
There's a way to tag people in the commit text:


> Co-authored-by: NAME <NAME@EXAMPLE.COM>

For me it would be:

> Co-authored-by: David Kraeutmann <kane@kane.cx>

**Comment by abigailalice** [comment]
Oh, I think `Fixed n` and `Char` also have a bug, due to not handling modifiers correctly. I'll work on that, as well as add the co-author disclaimer.

The failed build seems to just be from an unrelated test which happens to fail non-determinstically, using too precise rational comparison. I could fix that too, though I'm not sure if the 1e-15 is supposed to be correct, or just picked arbitrarily. I've noticed a few non-deterministic tests while working on this.

**Comment by abigailalice** [comment]
Okay, I fixed the bugs for those corner cases, added some test cases to cover them, and added proper attribution. My only major concern now is that I needed to force push to correct the first commit message, and that may have affected other commits in the middle. I'm newish to git in groups, so let me know if that's an issue, and I can just make another branch without the issue, and issue a new pull request. Otherwise, unless any of the bikeshedding is cause for concern, I'm happy with the current state of things.

I should say I've only tested this superficially on a live project; it was able to detect some bugs in my code in some rarely inserted into tables (due to fields which had been removed in Haskell, but not in the database). But otherwise most of the testing is from the test cases themselves.

**Comment by ocharles** [comment]
Thanks for all your hard work here @abigailalice and @KaneTW! There's quite a lot here so it might take me a while to get through this as I'm really busy at work at the moment, but I will try and get to this :pray:

**Comment by abigailalice** [comment]
Take your time, just remember that the middle commits aren't anything related to the actual PR, only the first and last are.

**Comment by ocharles** [comment]
> Just remember that the middle commits aren't anything related to the actual PR, only the first and last are.

This makes it very difficult to review. Could you rebase your branch so the changes are on master? Don't worry too much about commit attribution - we squash merge here anyway, so we can sort that separately.

**Comment by abigailalice** [comment]
Okay @ocharles, I think it should be at an acceptable state to review. Sorry about the difficulty.

**Comment by abigailalice** [comment]
@ocharles, is there something preventing this from being accepted, or have you just not had time to get around to it? Not an issue if it's the latter, just wanted to make sure if I misunderstood anything.

**Comment by ocharles** [comment]
@abigailalice purely time constraints I'm afraid! It's still on my radar. I'll be on holiday in a few days so this might need to wait until mid September

File: src/Rel8/Table/Verify.hs [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+-- In the event multiple columns have the same name, this will fail silently. To
+-- handle that case, see @'checkedShowCreateTable'@
+showCreateTable :: Rel8able k => TableSchema (k Name) -> String
+showCreateTable schema = showCreateTable_helper schema.name.name $ fmap NonEmpty.head $ schemaToTypeMap schema.columns
+
+-- |@'checkedShowCreateTable'@ shows an example CREATE TABLE statement for the
+-- table. This does not show relationships like primary or foreign keys, but can
+-- still be useful to see what types rel8 will expect of the underlying database.
+--
+-- In the event multiple columns have the same name, this will return a map of
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:357]:
```suggestion
-- In the event that multiple columns have the same name, this will return a map of
```

File: src/Rel8/Table/Verify.hs [hunk 2]
```diff
@@ -0,0 +1,642 @@
...
+-- names to the labels identifying the column.
+checkedShowCreateTable :: Rel8able k => TableSchema (k Name) -> Either (M.Map String (NonEmpty [String])) String
+checkedShowCreateTable schema = case checkedSchemaToTypeMap schema.columns of
+    Left e -> Left $ (fmap . fmap) (\typ -> typ.label) e
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:361]:
```suggestion
    Left e -> Left $ map (\(name, typ) -> (name, typ.label)) e
```
Slightly clearer

File: src/Rel8/Table/Verify.hs [hunk 3]
```diff
@@ -0,0 +1,642 @@
...
+    sameName = equalName db.typeName.name hs.typeName.name
+
+    toName :: TypeInfo -> String
+    toName typeInfo = case typeInfo.typeName.name of
+        QualifiedName name _ -> L.dropWhile (=='_') name
+
+equalName :: QualifiedName -> QualifiedName -> Bool
+equalName (QualifiedName a (Just b)) (QualifiedName a' (Just b'))
+  = L.dropWhile (=='_') a == L.dropWhile (=='_') a' && b == b'
+equalName (QualifiedName a _) (QualifiedName a' _)
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:390]:
```suggestion
equalName (QualifiedName a Nothing) (QualifiedName a' Nothing)
```
this is just a style nitpick

File: src/Rel8/Table/Verify.hs [hunk 4]
```diff
@@ -0,0 +1,642 @@
...
+    { tableExists :: Bool
+    , name :: String
+    , columns :: [ColumnInfo]
+    }
+  | DuplicateNames
+    { name :: String
+    , duplicates :: M.Map String (NonEmpty.NonEmpty TypeInfo)
+    }
+  deriving (Show)
+data ColumnInfo = ColumnInfo
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:510]:
```suggestion

data ColumnInfo = ColumnInfo
```

File: src/Rel8/Table/Verify.hs [hunk 5]
```diff
@@ -0,0 +1,642 @@
...
+    { name   :: String
+    , hsType :: Maybe TypeInfo
+    , dbType :: Maybe TypeInfo
+    , error :: Maybe ColumnError
+    } deriving (Show)
+data ColumnError
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:516]:
```suggestion

data ColumnError
```

File: src/Rel8/Table/Verify.hs [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+data SomeTableSchema where
+    -- The ToExpr constraint isn't used here, but can be used to read from the
+    -- SomeTableSchema, which can be useful to combine the type checking with more
+    -- thorough value-level checking of the validity of existing rows in the
+    -- table.
+    SomeTableSchema
+        :: (ToExprs (k Expr) (GFromExprs k), Rel8able k)
+        => TableSchema (k Name) -> SomeTableSchema
+
+-- |@'getSchemaErrors'@ checks whether the provided schemas have correct postgresql
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:590]:
```suggestion
-- |@'getSchemaErrors'@ checks whether the provided schemas have the correct PostgreSQL
```

File: src/Rel8/Table/Verify.hs [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+-- column names and types to allow reading and writing from their equivalent Haskell
+-- types, returning a list of errors if that is not the case. The function does not
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:592]:
```suggestion
-- types, and returns a list of errors if that is not the case. The function does not
```

File: src/Rel8/Table/Verify.hs [hunk 8]
```diff
@@ -0,0 +1,642 @@
...
+-- 3. no nonexistent columns are present in the hs type
+-- 4. no two columns in the same schema share the same name
+--
+-- It's still possible for a valid schema to allow invalid data, for instance,
+-- if using an ADT, which can introduce restrictions on which values are allowed
+-- for the column representing the tag, and introduce restrictions on which
+-- columns are non-null depending on the value of the tag. However, if the
+-- schema is valid rel8 shouldn't be able to write invalid data to the table.
+--
+-- However, it is possible for migrations to cause valid data to become invalid,
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:607]:
```suggestion
-- However, it is possible for migrations to cause valid data to become invalid
```

File: src/Rel8/Table/Verify.hs [hunk 9]
```diff
@@ -0,0 +1,642 @@
...
+-- in ways not detectable by this function, if the migration code changes the
+-- schema correctly but doesn't handle the value-level constraints correctly, so
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:609]:
```suggestion
-- schema correctly but doesn't handle the value-level constraints correctly. So
```

File: src/Rel8/Table/Verify.hs [hunk 10]
```diff
@@ -0,0 +1,642 @@
...
+-- it is a good idea to both read from the tables and check the schema for errors
+-- in a transaction during the migration. This former will catch value-level
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:611]:
```suggestion
-- in a transaction during the migration. The former will catch value-level
```

File: src/Rel8/Table/Verify.hs [hunk 11]
```diff
@@ -0,0 +1,642 @@
...
+-- userTable :: TableSchema (User Name)
+-- orderTable :: TableSchema (Order Name)
+--
+-- tables :: [SomeTableSchema]
+-- tables = [SomeTableSchema userTable, SomeTable orderTable]
+-- @
+--
+-- This is used by @'schemaErrors'@ to more conveniently group every table an
+-- application relies on together, for typechecking the postgresql schemas
+-- together in a single batch.
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:580]:
```suggestion
-- This is used by @'schemaErrors'@ to conveniently group every table an
-- application relies on for typechecking the PostgreSQL schema
-- together in a single batch.
```

File: src/Rel8/Table/Verify.hs [hunk 12]
```diff
@@ -0,0 +1,642 @@
...
+    -- thorough value-level checking of the validity of existing rows in the
+    -- table.
+    SomeTableSchema
+        :: (ToExprs (k Expr) (GFromExprs k), Rel8able k)
+        => TableSchema (k Name) -> SomeTableSchema
+
+-- |@'getSchemaErrors'@ checks whether the provided schemas have correct postgresql
+-- column names and types to allow reading and writing from their equivalent Haskell
+-- types, returning a list of errors if that is not the case. The function does not
+-- crash on an encountered bug, instead leaving it to the caller to decide how
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:593]:
```suggestion
-- crash on encountering a bug, instead leaving it to the caller to decide how
```

File: src/Rel8/Table/Verify.hs [hunk 13]
```diff
@@ -0,0 +1,642 @@
...
+-- schema correctly but doesn't handle the value-level constraints correctly, so
+-- it is a good idea to both read from the tables and check the schema for errors
+-- in a transaction during the migration. This former will catch value-level
+-- bugs, while the latter will help ensure the schema is set up correctly to
+-- be able to insert new data.
+--
+-- This function does nothing to check that the conflict target of an @Upsert@
+-- are valid for the schema, nor can it prevent invalid uses of @unsafeDefault@.
+-- However, it should be enough to catch the most likely errors.
+getSchemaErrors :: [SomeTableSchema] -> HS.Statement () (Maybe Text)
```
Comment (TeofilC) [inline src/Rel8/Table/Verify.hs:618]: Do we want Text or String here. Some of the other functions return String but this returns Text. I don't think it makes much of a difference but I wanted to check whether you wanted it to be consistent.

File: src/Rel8/Table/Verify.hs (hunk 3 above)
Comment (TeofilC) [reply src/Rel8/Table/Verify.hs:390]: Nope that's wrong oops. I'll drop this
